File: cores/typescript/src/cpu.ts
// 6502 CPU core (port of cores/c/src/cpu.c).

import { Bus } from './bus';

export enum AddrMode {
    Implied,
    Accumulator,
    Immediate,
    ZeroPage,
    ZeroPageX,
    ZeroPageY,
    Absolute,
    AbsoluteX,
    AbsoluteY,
    Indirect,
    IndirectX,
    IndirectY,
    Relative,
}

export enum Dummy {
    None,
    Read,
    Rmw,
}

export interface Operand {
    tag: number; // 0=None, 1=Acc, 2=Addr
    address: number;
}

export function makeNone(): Operand { return { tag: 0, address: 0 }; }
export function makeAcc(): Operand { return { tag: 1, address: 0 }; }
export function makeAddr(a: number): Operand { return { tag: 2, address: a }; }

export class Cpu {
    static readonly FlagC = 0x01;
    static readonly FlagZ = 0x02;
    static readonly FlagI = 0x04;
    static readonly FlagD = 0x08;
    static readonly FlagB = 0x10;
    static readonly FlagU = 0x20;
    static readonly FlagV = 0x40;
    static readonly FlagN = 0x80;

    static readonly VectorNmi = 0xFFFA;
    static readonly VectorReset = 0xFFFC;
    static readonly VectorIrq = 0xFFFE;

    static readonly NmiPending = 0x01;
    static readonly IrqPending = 0x02;
    static readonly Halted = 0x04;

    a: number = 0;
    x: number = 0;
    y: number = 0;
    sp: number = 0;
    pc: number = 0;
    status: number = 0;
    flags: number = 0;

    bus!: Bus;

    init(): void {
        this.a = 0; this.x = 0; this.y = 0;
        this.sp = 0xFD;
        this.pc = 0;
        this.status = Cpu.FlagU | Cpu.FlagI;
        this.flags = 0;
    }

    carry(): boolean { return (this.status & Cpu.FlagC) !== 0; }
    zero(): boolean { return (this.status & Cpu.FlagZ) !== 0; }
    interruptDisable(): boolean { return (this.status & Cpu.FlagI) !== 0; }
    decimal(): boolean { return (this.status & Cpu.FlagD) !== 0; }
    overflow(): boolean { return (this.status & Cpu.FlagV) !== 0; }
    negative(): boolean { return (this.status & Cpu.FlagN) !== 0; }

    setFlag(flag: number, v: boolean): void {
        if (v) this.status |= flag;
        else this.status &= (flag ^ 0xFF);
    }

    setCarry(v: boolean): void { this.setFlag(Cpu.FlagC, v); }
    setZero(v: boolean): void { this.setFlag(Cpu.FlagZ, v); }
    setInterruptDisable(v: boolean): void { this.setFlag(Cpu.FlagI, v); }
    setDecimal(v: boolean): void { this.setFlag(Cpu.FlagD, v); }
    setOverflow(v: boolean): void { this.setFlag(Cpu.FlagV, v); }
    setNegative(v: boolean): void { this.setFlag(Cpu.FlagN, v); }

    setNz(value: number): void {
        this.setZero(value === 0);
        this.setNegative((value & 0x80) !== 0);
    }

    fetchByte(): number {
        const b = this.bus.read(this.pc);
        this.pc = (this.pc + 1) & 0xFFFF;
        return b;
    }

    fetchWord(): number {
        const lo = this.fetchByte();
        const hi = this.fetchByte();
        return (lo | (hi << 8)) & 0xFFFF;
    }

    push(value: number): void {
        this.bus.write((0x0100 | this.sp) & 0xFFFF, value);
        this.sp = (this.sp - 1) & 0xFF;
    }

    pull(): number {
        this.sp = (this.sp + 1) & 0xFF;
        return this.bus.read((0x0100 | this.sp) & 0xFFFF);
    }

    pushPc(pc: number): void {
        this.push((pc >> 8) & 0xFF);
        this.push(pc & 0xFF);
    }

    pullPc(): number {
        const lo = this.pull();
        const hi = this.pull();
        return (lo | (hi << 8)) & 0xFFFF;
    }

    pushStatus(withBreak: boolean): void {
        let p = this.status | Cpu.FlagU;
        if (withBreak) p |= Cpu.FlagB;
        else p &= (Cpu.FlagB ^ 0xFF);
        this.push(p);
    }

    pullStatus(): void {
        const p = this.pull();
        this.status = (p & (Cpu.FlagB ^ 0xFF)) | Cpu.FlagU;
    }

    readOperand(op: Operand): number {
        switch (op.tag) {
            case 1: return this.a;
            case 2: return this.bus.read(op.address);
            default: return 0;
        }
    }

    writeOperand(op: Operand, value: number): void {
        switch (op.tag) {
            case 1: this.a = value; return;
            case 2: this.bus.write(op.address, value); return;
        }
    }

    private serviceInterrupt(vector: number): void {
        this.pushPc(this.pc);
        this.pushStatus(false);
        this.setInterruptDisable(true);
        this.pc = this.readVector(vector);
    }

    nmi(): void { this.serviceInterrupt(Cpu.VectorNmi); }
    irq(): void { this.serviceInterrupt(Cpu.VectorIrq); }

    reset(): void {
        this.sp = 0xFD;
        this.setInterruptDisable(true);
        this.status |= Cpu.FlagU;
        this.pc = this.readVector(Cpu.VectorReset);
        this.flags &= (Cpu.Halted ^ 0xFF);
    }

    step(): number {
        if ((this.flags & Cpu.Halted) !== 0) return 1;
        if ((this.flags & Cpu.NmiPending) !== 0) {
            this.flags &= (Cpu.NmiPending ^ 0xFF);
            this.nmi();
            return 7;
        }
        if ((this.flags & Cpu.IrqPending) !== 0 && !this.interruptDisable()) {
            this.flags &= (Cpu.IrqPending ^ 0xFF);
            this.irq();
            return 7;
        }
        const opcode = this.fetchByte();
        return this.execute(opcode);
    }

    isHalted(): boolean { return (this.flags & Cpu.Halted) !== 0; }
    setHalted(v: boolean): void {
        if (v) this.flags |= Cpu.Halted;
        else this.flags &= (Cpu.Halted ^ 0xFF);
    }
    nmiPendingFlag(): boolean { return (this.flags & Cpu.NmiPending) !== 0; }
    irqPendingFlag(): boolean { return (this.flags & Cpu.IrqPending) !== 0; }
    setNmiPending(v: boolean): void {
        if (v) this.flags |= Cpu.NmiPending;
        else this.flags &= (Cpu.NmiPending ^ 0xFF);
    }
    setIrqPending(v: boolean): void {
        if (v) this.flags |= Cpu.IrqPending;
        else this.flags &= (Cpu.IrqPending ^ 0xFF);
    }

    readVector(addr: number): number {
        const lo = this.bus.read(addr);
        const hi = this.bus.read((addr + 1) & 0xFFFF);
        return (lo | (hi << 8)) & 0xFFFF;
    }
}

File: cores/csharp/src/Cpu.cs
using System.Runtime.CompilerServices;

namespace NesCore;

// 6502 addressing modes (addressing.rs AddrMode).
public enum AddrMode : byte
{
    Implied,
    Accumulator,
    Immediate,
    ZeroPage,
    ZeroPageX,
    ZeroPageY,
    Absolute,
    AbsoluteX,
    AbsoluteY,
    Indirect,
    IndirectX,
    IndirectY,
    Relative,
}

// Dummy-read behaviour (addressing.rs Dummy).
public enum Dummy : byte
{
    None,
    Read,
    Rmw,
}

// Operand tag (addressing.rs Operand).
public struct Operand
{
    public const int None = 0;
    public const int Acc = 1;
    public const int Addr = 2;

    public int Tag;
    public ushort Address;

    public static Operand MakeNone() => new Operand { Tag = None };
    public static Operand MakeAcc() => new Operand { Tag = Acc };
    public static Operand MakeAddr(ushort a) => new Operand { Tag = Addr, Address = a };
}

// 6502 CPU core (mod.rs struct Cpu).
public sealed partial class Cpu
{
    public const byte FlagC = 0x01;
    public const byte FlagZ = 0x02;
    public const byte FlagI = 0x04;
    public const byte FlagD = 0x08;
    public const byte FlagB = 0x10;
    public const byte FlagU = 0x20;
    public const byte FlagV = 0x40;
    public const byte FlagN = 0x80;

    public const ushort VectorNmi = 0xFFFA;
    public const ushort VectorReset = 0xFFFC;
    public const ushort VectorIrq = 0xFFFE;

    public const byte NmiPending = 0x01;
    public const byte IrqPending = 0x02;
    public const byte Halted = 0x04;

    public byte A;
    public byte X;
    public byte Y;
    public byte Sp;
    public ushort Pc;
    public byte Status;
    public byte Flags;

    public Bus Bus;

    public Cpu() { }

    public void Init()
    {
        A = 0; X = 0; Y = 0;
        Sp = 0xFD;
        Pc = 0;
        Status = FlagU | FlagI;
        Flags = 0;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public bool Carry() => (Status & FlagC) != 0;
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public bool Zero() => (Status & FlagZ) != 0;
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public bool InterruptDisable() => (Status & FlagI) != 0;
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public bool Decimal() => (Status & FlagD) != 0;
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public bool Overflow() => (Status & FlagV) != 0;
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public bool Negative() => (Status & FlagN) != 0;

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public void SetFlag(byte flag, bool v)
    {
        if (v) Status |= flag;
        else Status &= (byte)(flag ^ 0xFF);
    }

    public void SetCarry(bool v) => SetFlag(FlagC, v);
    public void SetZero(bool v) => SetFlag(FlagZ, v);
    public void SetInterruptDisable(bool v) => SetFlag(FlagI, v);
    public void SetDecimal(bool v) => SetFlag(FlagD, v);
    public void SetOverflow(bool v) => SetFlag(FlagV, v);
    public void SetNegative(bool v) => SetFlag(FlagN, v);

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public void SetNz(byte value)
    {
        SetZero(value == 0);
        SetNegative((value & 0x80) != 0);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public byte FetchByte()
    {
        byte b = Bus.Read(Pc);
        Pc++;
        return b;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public ushort FetchWord()
    {
        ushort lo = FetchByte();
        ushort hi = FetchByte();
        return (ushort)(lo | (hi << 8));
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public void Push(byte value)
    {
        Bus.Write((ushort)(0x0100 | Sp), value);
        Sp--;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public byte Pull()
    {
        Sp++;
        return Bus.Read((ushort)(0x0100 | Sp));
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public void PushPc(ushort pc)
    {
        Push((byte)(pc >> 8));
        Push((byte)(pc & 0xFF));
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public ushort PullPc()
    {
        ushort lo = Pull();
        ushort hi = Pull();
        return (ushort)(lo | (hi << 8));
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public void PushStatus(bool withBreak)
    {
        byte p = (byte)(Status | FlagU);
        if (withBreak) p |= FlagB;
        else p &= (byte)(FlagB ^ 0xFF);
        Push(p);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public void PullStatus()
    {
        byte p = Pull();
        Status = (byte)((p & (byte)(FlagB ^ 0xFF)) | FlagU);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public byte ReadOperand(Operand op)
    {
        switch (op.Tag)
        {
            case Operand.Acc: return A;
            case Operand.Addr: return Bus.Read(op.Address);
            default: return 0;
        }
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public void WriteOperand(Operand op, byte value)
    {
        switch (op.Tag)
        {
            case Operand.Acc: A = value; return;
            case Operand.Addr: Bus.Write(op.Address, value); return;
        }
    }

    private void ServiceInterrupt(ushort vector)
    {
        PushPc(Pc);
        PushStatus(false);
        SetInterruptDisable(true);
        Pc = ReadVector(vector);
    }

    public void Nmi() => ServiceInterrupt(VectorNmi);
    public void Irq() => ServiceInterrupt(VectorIrq);

    public void Reset()
    {
        Sp = 0xFD;
        SetInterruptDisable(true);
        Status |= FlagU;
        Pc = ReadVector(VectorReset);
        Flags &= (byte)(Halted ^ 0xFF);
    }

    public byte Step()
    {
        if ((Flags & Halted) != 0) return 1;
        if ((Flags & NmiPending) != 0)
        {
            Flags &= (byte)(NmiPending ^ 0xFF);
            Nmi();
            return 7;
        }
        if ((Flags & IrqPending) != 0 && !InterruptDisable())
        {
            Flags &= (byte)(IrqPending ^ 0xFF);
            Irq();
            return 7;
        }
        byte opcode = FetchByte();
        return Execute(opcode);
    }

    public bool IsHalted() => (Flags & Halted) != 0;
    public void SetHalted(bool v)
    {
        if (v) Flags |= Halted;
        else Flags &= (byte)(Halted ^ 0xFF);
    }
    public bool NmiPendingFlag() => (Flags & NmiPending) != 0;
    public bool IrqPendingFlag() => (Flags & IrqPending) != 0;
    public void SetNmiPending(bool v)
    {
        if (v) Flags |= NmiPending;
        else Flags &= (byte)(NmiPending ^ 0xFF);
    }
    public void SetIrqPending(bool v)
    {
        if (v) Flags |= IrqPending;
        else Flags &= (byte)(IrqPending ^ 0xFF);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public ushort ReadVector(ushort addr)
    {
        ushort lo = Bus.Read(addr);
        ushort hi = Bus.Read((ushort)(addr + 1));
        return (ushort)(lo | (hi << 8));
    }
}
